hydra_pywr/nodes/hydropower.py

```diff
@@ -52,157 +99,94 @@ def __init__(self, model, name, storage_node, release_values, scenario=None, **k
-        rainfall_cost = kwargs.pop('rainfall_cost', -999)
-        # Assume rainfall/evap is mm/day
-        # Need to convert:
-        #   Mm2 -> m2
-        #   mm/day -> m/day
-        #   m3/day -> Mm3/day
-        # TODO allow this to be configured
+        self.weather_cost = kwargs.pop('weather_cost', -999)
+        self.evaporation_cost = kwargs.pop('evaporation_cost', -999)
+        self.rainfall_cost = kwargs.pop('rainfall_cost', -999)
         const = kwargs.pop('const', 1e6 * 1e-3 * 1e-6)
 
-        super().__init__(model, name, **kwargs)
+        # Pywr Storage does not expect a 'weather' kwargs, so move this to instance
+        self.weather = kwargs.pop("weather", None)
 
-        self._set_bathymetry(model, bathymetry, volume, level, area)
+        super().__init__(model, name, **kwargs)
 
         self.const = ConstantParameter(model, const)
 
         self.rainfall_node = None
         self.rainfall_recorder = None
         self.evaporation_node = None
         self.evaporation_recorder = None
-        if weather is not None:
-            self._make_weather_nodes(model, weather, weather_cost)
-        else:
-            self._make_evaporation_node(model, evaporation, evaporation_cost)
-            self._make_rainfall_node(model, rainfall, rainfall_cost)
 
-    def _set_bathymetry(self, model, bathymetry, volume=None, level=None, area=None):
+
+    @classmethod
+    def pre_load(cls, model, data):
+
+        bathymetry = data.pop("bathymetry", None)
+        name = data.pop("name")
+        node = cls(name=name, model=model, **data)
+        volumes = None
+
         if bathymetry is not None:
-            bathymetry = pd.DataFrame.from_dict(bathymetry)
-            volumes = bathymetry['volume'].astype(np.float64)
-            levels = bathymetry['level'].astype(np.float64)
-            areas = bathymetry['area'].astype(np.float64)
-        elif volume is not None and level is not None and area is not None:
-            #cater for when they are not specified as parameters, eg they are
-            #scalars or standard dataframes
-            volumes = volume
-            levels = level
-            areas = area
-        else:
-            try:
-                volumes = load_parameter(model, f'__{self.name}__:volume')
-            except KeyError:
-                log.warning(f"Please specify a bathymetry or volume on node {self.name}")
-                volumes = None
-            try:
-                areas = load_parameter(model, f'__{self.name}__:area')
-            except KeyError:
-                log.warning(f"Please specify a bathymetry or area on node {self.name}")
-                areas = None
-            try:
-                levels = load_parameter(model, f'__{self.name}__:level')
-            except KeyError:
-                log.warning(f"Please specify a bathymetry or level on node {self.name}")
-                levels = None
+            if isinstance(bathymetry, str):
```

Comment:
Lots of try-excepts removed. Is this data guaranteed to be there then, if the checks are being removed?

Reply:
I think the case where there is no bathymetry but there is a separate value for each of volume, level and area needs some warnings restoring if one or more of these is not provided - will revise to add this.